### PR TITLE
[quantization] Store bias in PackedConvWeight in fbgemm

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25626 [quantization] Store bias in PackedConvWeight in fbgemm**

Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

### CODE REVIEW

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py [hunk 1]
```diff
@@ -24,29 +24,13 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                                          stride=stride, padding=padding, dilation=dilation,
                                          groups=groups, bias=bias, padding_mode=padding_mode)
 
-    def weight(self):
```
Comment (dskhudia) [inline torch/nn/_intrinsic/quantized/modules/conv_relu.py]: Are these two methods being removed?

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 2]
```diff
@@ -124,20 +123,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+            at::dequantize(bias), pack_ptr.w_scale[0] * act_scale, 0, kQInt32);
+      } else if (pack_ptr.q_scheme == kPerChannelAffine) {
+        std::vector<int64_t> vect(1, 0);
+        IntArrayRef axis(vect.data(), 1);
+        at::Tensor bias_scale = at::ones({K}, at::dtype(at::kDouble));
+        at::Tensor bias_zp = at::zeros({K}, at::dtype(at::kLong));
+        for (int i = 0; i < K; ++i) {
+          bias_scale.data_ptr<double>()[i] = pack_ptr.w_scale[i] * act_scale;
+        }
+        qbias = quantize_linear_per_channel_cpu(
```
[the remark this answers is not in the document]
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Using the `quantize_linear_per_channel` function causes the following error 
```
RuntimeError: Expected object of type Variable but found type CPUFloatType for argument #1 'scales' (checked_cast_variable at caffe2/torch/csrc/autograd/VariableTypeManual.cpp:38)
```
Spoke to @yf225 offline and he said the error suggests that we are missing a at::AutoNonVariableTypeMode thread-local guard somewhere in the call path of QLinearInt8 operator().
quantize_linear_per_channel dispatches to VariableType dispatcher, which expects the input tensor to be a Variable.

[the remark this answers is not in the document]
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: do you know why are we missing this guard?

[the remark this answers is not in the document]
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: I am not sure. @smessmer any thoughts on if this guard should be added?

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py (hunk 1 above)
Comment (supriyar) [reply torch/nn/_intrinsic/quantized/modules/conv_relu.py]: Yes, I don't see why they are needed as the parent module already defined them. Looks like linear_relu.py also doesn't have them

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 3]
```diff
@@ -124,20 +123,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+    at::Tensor qbias;
+    if (pack_ptr.bias.has_value()) {
+      at::Tensor bias = pack_ptr.bias.value();
+      // Temporary: Quantize bias
+      qbias = bias;
+      if (pack_ptr.q_scheme == kPerTensorAffine) {
+        qbias = at::quantize_linear(
+            at::dequantize(bias), pack_ptr.w_scale[0] * act_scale, 0, kQInt32);
+      } else if (pack_ptr.q_scheme == kPerChannelAffine) {
+        std::vector<int64_t> vect(1, 0);
```
[the remark this answers is not in the document]
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Done.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (raghuramank100) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Is this issue resolved?

File: test/test_quantized.py [hunk 4]
```diff
@@ -1136,9 +1135,9 @@ def test_qconv_unpack(self, X, strideH, strideW, padH, padW, channelwise):
         strides = [strideH, strideW]
         paddings = [padH, padW]
         dilations = [1, 1]
-        W_packed = qconv_prepack(W_q, strides, paddings, dilations, groups)
+        W_packed = qconv_prepack(W_q, None, strides, paddings, dilations, groups)
         # Unpack weights weight unpacking operator (Used for serialization)
-        W_unpacked = qconv_unpack(W_packed)
+        W_unpacked = qconv_unpack(W_packed)[0]
```
Comment (raghuramank100) [inline test/test_quantized.py]: Please add a check to ensure that the unpacked bias is identical to the original bias.

**Review by raghuramank100 — APPROVED** [review]
Looks great, please address comments prior to submitting.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Not yet, I will file a separate issue for this.

**Pull request closed without merging**